File: Dsa-day66.cpp
#include <bits/stdc++.h>
using namespace std;
class MovieRentingSystem {
    unordered_map<int, set<pair<int,int>>> movieMap;
    set<tuple<int,int,int>> rented;
    unordered_map<long long,int> priceMap;
    long long key(int shop,int movie){ return ((long long)shop<<32) | (unsigned long long)movie; }
public:
    MovieRentingSystem(int n, vector<vector<int>>& entries) {
        for(auto &e: entries){
            int shop=e[0], movie=e[1], price=e[2];
            movieMap[movie].insert({price,shop});
            priceMap[key(shop,movie)] = price;
        }
    }
    vector<int> search(int movie) {
        vector<int> res;
        auto it = movieMap.find(movie);
        if(it==movieMap.end()) return res;
        int cnt=0;
        for(auto &p: it->second){
            res.push_back(p.second);
            if(++cnt==5) break;
        }
        return res;
    }
    void rent(int shop, int movie) {
        int price = priceMap[key(shop,movie)];
        movieMap[movie].erase({price,shop});
        rented.insert({price,shop,movie});
    }
    void drop(int shop, int movie) {
        int price = priceMap[key(shop,movie)];
        movieMap[movie].insert({price,shop});
        rented.erase({price,shop,movie});
    }
    vector<vector<int>> report() {
        vector<vector<int>> res;
        int cnt=0;
        for(auto &t: rented){
            if(cnt==5) break;
            int price,getshop,getmovie;
            tie(price,getshop,getmovie) = t;
            res.push_back({getshop,getmovie});
            ++cnt;
        }
        return res;
    }
};

File: Dsa-day66.java
import java.util.*;
class MovieRentingSystem {
    static class Pair {
        int price, shop;
        Pair(int p,int s){price=p;shop=s;}
    }
    static class Triple {
        int price, shop, movie;
        Triple(int p,int s,int m){price=p;shop=s;movie=m;}
    }
    Map<Integer, TreeSet<Pair>> movieMap;
    TreeSet<Triple> rented;
    Map<Long, Integer> priceMap;
    Comparator<Pair> pairComp = (a,b)-> a.price!=b.price ? Integer.compare(a.price,b.price) : Integer.compare(a.shop,b.shop);
    Comparator<Triple> tripleComp = (a,b)-> a.price!=b.price ? Integer.compare(a.price,b.price) : (a.shop!=b.shop?Integer.compare(a.shop,b.shop):Integer.compare(a.movie,b.movie));
    public MovieRentingSystem(int n, int[][] entries) {
        movieMap = new HashMap<>();
        rented = new TreeSet<>(tripleComp);
        priceMap = new HashMap<>();
        for(int[] e: entries){
            int shop=e[0], movie=e[1], price=e[2];
            movieMap.computeIfAbsent(movie, k->new TreeSet<>(pairComp)).add(new Pair(price,shop));
            priceMap.put(key(shop,movie), price);
        }
    }
    long key(int shop,int movie){return (((long)shop)<<32) | (movie & 0xffffffffL);}
    public List<Integer> search(int movie) {
        List<Integer> res = new ArrayList<>();
        TreeSet<Pair> set = movieMap.get(movie);
        if(set==null || set.isEmpty()) return res;
        Iterator<Pair> it = set.iterator();
        int k=0;
        while(it.hasNext() && k<5){
            res.add(it.next().shop);
            k++;
        }
        return res;
    }
    public void rent(int shop, int movie) {
        int price = priceMap.get(key(shop,movie));
        TreeSet<Pair> set = movieMap.get(movie);
        set.remove(new Pair(price,shop));
        rented.add(new Triple(price,shop,movie));
    }
    public void drop(int shop, int movie) {
        int price = priceMap.get(key(shop,movie));
        TreeSet<Pair> set = movieMap.computeIfAbsent(movie, k->new TreeSet<>(pairComp));
        set.add(new Pair(price,shop));
        rented.remove(new Triple(price,shop,movie));
    }
    public List<List<Integer>> report() {
        List<List<Integer>> res = new ArrayList<>();
        Iterator<Triple> it = rented.iterator();
        int k=0;
        while(it.hasNext() && k<5){
            Triple t = it.next();
            res.add(Arrays.asList(t.shop, t.movie));
            k++;
        }
        return res;
    }
}
